include/ofi_coll.h

```diff
@@ -104,52 +109,67 @@ struct util_coll_reduce_item {
 	enum fi_op		op;
 };
 
-struct util_coll_comp_item;
-
-typedef void (*util_coll_comp_t)(struct util_coll_mc *coll_mc,
-				 struct util_coll_comp_item *comp);
-
-struct util_coll_join_comp_data {
-	uint64_t		cid_buf[OFI_CONTEXT_ID_SIZE];
-	uint64_t		tmp_cid_buf[OFI_CONTEXT_ID_SIZE];
-};
-
-struct util_coll_comp_item {
-	struct util_coll_hdr	hdr;
-	enum util_coll_op_type	op_type;
-	void			*data;
-	util_coll_comp_t	comp_fn;
-};
-
 struct util_coll_mc {
 	struct fid_mc		mc_fid;
 	struct fid_ep		*ep;
-	struct util_av_set 	*av_set;
+	struct util_av_set	*av_set;
 	struct slist		barrier_list;
 	struct slist		deferred_list;
-	int 			my_rank;
-	uint16_t		cid;
-	uint16_t		tag_seq;
+	struct slist		pending_xfer_list;
+	int			my_rank;
+	uint16_t		cid_id;
+	uint16_t		cid_seq;
 	ofi_atomic32_t		ref;
 };
 
+struct util_coll_state;
+
+typedef void (*util_coll_comp_fn_t)(struct util_coll_state *state);
+
+struct util_coll_cid_data {
+	uint64_t	cid_buf[OFI_CONTEXT_ID_SIZE];
+	uint64_t	tmp_cid_buf[OFI_CONTEXT_ID_SIZE];
+};
+
+struct util_coll_state {
+	struct util_coll_hdr	hdr;
+	struct dlist_entry	entry;
+	enum util_coll_op_type	type;
+	void			*context;
+	struct util_coll_mc	*mc;
+	union data {
+		struct util_coll_cid_data join;
+		struct util_coll_cid_data barrier;
+	} data;
+	util_coll_comp_fn_t	comp_fn;
+};
+
 int ofi_join_collective(struct fid_ep *ep, fi_addr_t coll_addr,
 			const struct fid_av_set *set, uint64_t flags,
 			struct fid_mc **mc, void *context);
 
 int ofi_av_set(struct fid_av *av, struct fi_av_set_attr *attr,
-	       struct fid_av_set **av_set_fid, void * context);
+	       struct fid_av_set **av_set_fid, void *context);
 
 ssize_t ofi_ep_barrier(struct fid_ep *ep, fi_addr_t coll_addr, void *context);
 
-ssize_t ofi_ep_writeread(struct fid_ep *ep, const void *buf, size_t count,
-		     void *desc, void *result, void *result_desc,
-		     fi_addr_t coll_addr, enum fi_datatype datatype,
-		     enum fi_op op, uint64_t flags, void *context);
+ssize_t ofi_ep_no_barrier(struct fid_ep *ep, fi_addr_t coll_addr, void *context);
+
+ssize_t ofi_ep_no_writeread(struct fid_ep *ep, const void *buf, size_t count,
+			 void *desc, void *result, void *result_desc,
+			 fi_addr_t coll_addr, enum fi_datatype datatype,
+			 enum fi_op op, uint64_t flags, void *context);
+
+ssize_t ofi_ep_no_writereadmsg(struct fid_ep *ep,
+			    const struct fi_msg_collective *msg,
+			    struct fi_ioc *resultv, void **result_desc,
+			    size_t result_count, uint64_t flags);
```

Comment:
The 'no' calls should be defined in include/ofi_enosys.h, with the implementations in src/enosys.c.